blog/2016-04-20-futhark-and-pygame.rst

```diff
@@ -0,0 +1,215 @@
+transform the returned NumPy array into something PyGame can
+understand, after which we blit it to the screen::
+
+  def render():
+      frame = l.render_frame(history)
+      pygame.surfarray.blit_array(surface, frame)
+      screen.blit(surface, (0, 0))
+      pygame.display.flip()
+
+Here, the ``l`` variable is the imported simulation backend,
+i.e. either the default ``life`` or one of the alternative variants
+``quadlife`` or ``quadlife_alt``, whose rules are invented by `Torben
+Mogensen <http://www.diku.dk/~torbenm/>`_.  When running the
+visualisation, you can specify the backend with the ``--variant``
+command-line flag.
+  
+Finally, since Game of Life is a state-based simulation, we need a way
+to step through the simulation, using previous output as new input.
+This is pretty simple in Python::
+
+  while True:
+      world, history = l.steps(world, history, steps)
+      render()
+      for event in pygame.event.get():
+          if event.type == pygame.QUIT:
+              sys.exit()
+
+The ``steps`` argument is the number of simulation steps to perform
+per frame, and defaults to 3.  You can set this to any positive 32-bit
+int.  To increase the work done per frame, the default is 3 and not 1.
+
+We have also added a simple PyGame event check, so that you can close
+the simulation window as expected.
+
+
+The Three Other Ones
+--------------------
+
+In the fluid simulator you can add both particles and forces.  See for yourself:
+
+.. raw:: html
+
+   <video src="/static/fluid-2016.04.20.webm"
+          poster="/static/fluid-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/fluid-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/fluid-2016.04.20.webm">fluid-2016.04.20.webm</a></p>
+
+My GPU is not the best one around, so I am running this in a fairly
+small window.
+
+The Mandelbrot Explorer is also pretty nifty:
+
+.. raw:: html
+
+   <video src="/static/mandelbrot-2016.04.20.webm"
+          poster="/static/mandelbrot-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/mandelbrot-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/mandelbrot-2016.04.20.webm">mandelbrot-2016.04.20.webm</a></p>
+
+In the end of the video, I switch to NumPy's internal Mandelbrot generator.
+
+Finally, there is the HotSpot 2D Heat Equation GUI.  You can see its
+performance `here </performance.html#hotspot-futhark-rodinia>`_.  This
+visualisation is pretty silly.
```

Comment:
But it looks kind of glowy and pretty!  The mess during startup is because the iniitial heat levels are random (but high), and it takes a while to dissipate.